staticd/static_nb_config.c

```diff
@@ -0,0 +1,1245 @@
+	int nh_type;
+
+	nh_type = yang_dnode_get_enum(args->dnode, "./nh-type");
+	ifname = yang_dnode_get_string(args->dnode, "./interface");
+	nh_vrf = yang_dnode_get_string(args->dnode, "./vrf");
+
+	nh = nb_running_get_entry(args->dnode, NULL, true);
+
+	pn_dnode = yang_dnode_get_parent(args->dnode, "path-list");
+	pn = nb_running_get_entry(pn_dnode, NULL, true);
+
+	src_dnode = yang_dnode_get_parent(pn_dnode, "src-list");
+	src_rn = nb_running_get_entry(src_dnode, NULL, true);
+
+	rn_dnode = yang_dnode_get_parent(src_dnode, "route-list");
+	rn = nb_running_get_entry(rn_dnode, NULL, true);
+	info = route_table_get_info(rn->table);
+
+	static_install_nexthop(src_rn, pn, nh, info->safi, info->svrf, ifname,
+			       nh_type, nh_vrf);
+}
+int routing_control_plane_protocols_control_plane_protocol_staticd_route_list_path_list_frr_nexthops_nexthop_mpls_label_stack_entry_pre_validate(
+	struct nb_cb_pre_validate_args *args)
+{
+	const struct lyd_node *mls_dnode;
+	uint32_t count;
+
+	mls_dnode = yang_dnode_get(args->dnode, "./mpls-label-stack");
+	count = yang_get_list_elements_count(yang_dnode_get_child(mls_dnode));
+
+	if (count > MPLS_MAX_LABELS) {
+		flog_warn(EC_LIB_NB_CB_CONFIG_VALIDATE,
+			  "Too many labels, Enter %d or fewer",
+			  MPLS_MAX_LABELS);
+		return NB_ERR_VALIDATION;
+	}
+	return NB_OK;
+}
+
+int routing_control_plane_protocols_name_validate(
+	struct nb_cb_create_args *args)
+{
+	const char *name;
+	name = yang_dnode_get_string(args->dnode, "./name");
+	if (!strmatch(name, "staticd")) {
+		flog_warn(
+			EC_LIB_NB_CB_CONFIG_VALIDATE,
+			"static routing supports only one instance with name staticd");
+		return NB_ERR_VALIDATION;
+	}
+	return NB_OK;
+}
+/*
+ * XPath:
+ * /frr-routing:routing/control-plane-protocols/control-plane-protocol/frr-staticd:staticd/route-list
+ */
+int routing_control_plane_protocols_control_plane_protocol_staticd_route_list_create(
+	struct nb_cb_create_args *args)
+{
+	struct vrf *vrf;
+	struct static_vrf *s_vrf;
+	struct route_node *rn;
+	const struct lyd_node *vrf_dnode;
+	struct prefix prefix;
+	const char *vrfname;
+	const char *afi_safi;
+	afi_t prefix_afi;
+	afi_t afi;
+	safi_t safi;
+
+	switch (args->event) {
+	case NB_EV_VALIDATE:
+		yang_dnode_get_prefix(&prefix, args->dnode, "./prefix");
+		afi_safi = yang_dnode_get_string(args->dnode, "./afi-safi");
+		yang_afi_safi_identity2value(afi_safi, &afi, &safi);
+		prefix_afi = family2afi(prefix.family);
+		if (afi != prefix_afi) {
+			flog_warn(
+				EC_LIB_NB_CB_CONFIG_VALIDATE,
+				"route node %s creation failed",
+				yang_dnode_get_string(args->dnode, "./prefix"));
+			return NB_ERR_VALIDATION;
+		}
+		break;
+	case NB_EV_PREPARE:
+	case NB_EV_ABORT:
+		break;
+	case NB_EV_APPLY:
+		vrf_dnode = yang_dnode_get_parent(args->dnode,
+						  "control-plane-protocol");
+		vrfname = yang_dnode_get_string(vrf_dnode, "./vrf");
+		vrf = vrf_lookup_by_name(vrfname);
+		vrf = vrf ? vrf : vrf_get(VRF_UNKNOWN, vrfname);
+		if (!vrf) {
```

Comment:
Shouldn't we create a pre-provisioned VRF on `routing_control_plane_protocols_control_plane_protocol_create()` instead of doing only a VRF lookup there? If we don't do that, these checks will need to be replicated across all routing daemons long term. Ideally a simple `nb_running_get_entry()` should be enough the get the VRF.

Reply:
done